x/slashing/keeper/infractions.go

```diff
@@ -58,19 +58,29 @@ func (k Keeper) HandleValidatorSignatureWithParams(ctx context.Context, params t
 	signedBlocksWindow := params.SignedBlocksWindow
 
 	// Compute the relative index, so we count the blocks the validator *should*
-	// have signed. We will use the 0-value default signing info if not present,
-	// except for start height. The index is in the range [0, SignedBlocksWindow)
+	// have signed. We will also use the 0-value default signing info if not present.
+	// The index is in the range [0, SignedBlocksWindow)
 	// and is used to see if a validator signed a block at the given height, which
 	// is represented by a bit in the bitmap.
-	index := signInfo.IndexOffset % signedBlocksWindow
-	signInfo.IndexOffset++
+	// The validator start height should get mapped to index 0, so we computed index as:
+	// (height - startHeight) % signedBlocksWindow
+	//
+	// NOTE: There is subtle different behavior between genesis validators and non-genesis validators.
+	// A genesis validator will start at index 0, whereas a non-genesis validator's startHeight will be the block
+	// they bonded on, but the first block they vote on will be one later. (And thus their first vote is at index 1)
```

Comment:
This is safe / should pose no issues imo.

We only look for liveness slashes after a full SignedBlockWindow blocks have passed. This has always granted one extra block for genesis validators, and exactly `SignedBlockWindow` number of blocks for non-genesis validators, before liveness slashing is looked at.

During a migration, this would cause the liveness tracking for the non-genesis validators to "skip" one block ahead. E.g. suppose the upgrade is at block N.

Pre-upgrade (block N-1) is at index 5. Post-upgrade (block N), my liveness index is 7 (not 6). So we "inherit" one old block in the liveness tracking for this window, but it gets fully fixed a SignedBlockWindow later. I really don't think this is an issue, and for genesis validators the index is preserved correctly.

If anything, I think this is generally lowering the complexity of the logic difference for the genesis/non-genesis discrepancy.